### PR TITLE
Fix recompute checkpointing + training CGs

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/tensor_parallel/random.py [hunk 1]
```diff
@@ -639,6 +639,13 @@ def checkpoint(
 ) -> _R:
     """Checkpoint a model or part of the model.
     This has been directly copied from torch.utils.checkpoint."""
+    from megatron.core.transformer.cuda_graphs import is_graph_capturing, is_graph_warmup
+
+    # Skip checkpointing during CUDA graph warmup and capture, matching the behavior of
+    # CheckpointWithoutOutput. The graph captures all ops directly; recomputation cannot
+    # run inside a captured graph.
```
[the remark this answers is not in the document]
Comment (tdene) [reply megatron/core/tensor_parallel/random.py:646]: You're too verbose, Claude.

File: megatron/core/tensor_parallel/random.py [hunk 2]
```diff
@@ -639,6 +639,13 @@ def checkpoint(
...
+    if is_graph_warmup() or is_graph_capturing():
```
Comment (jiemingz) [inline megatron/core/tensor_parallel/random.py:647]: Doesnt this just disable recomputation during graph capture? `is_graph_capturing()` is true when capturing. Why cant recomputation work with graphs?

Comment (tdene) [reply megatron/core/tensor_parallel/random.py:647]: Right, this only disables recomputation during graph capture. That's where I'm seeing errors in `main` right now: the first time we run through a training pass, we error out because we are capturing the graph while also attempting recompute.

Subsequent training passes seem to run fine.

**Review by jiemingz — APPROVED** [review]
There might be a way to get recompute working with per layer graphs but I am okay with explicitly disabling it like this unless needed.

**Review by ericharper — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
